rasa/shared/core/domain.py

```diff
@@ -320,127 +258,17 @@ def merge_domain_dicts(
-                combined.get(key, []), domain_dict.get(key, [])
-            )
-            combined[key] = self.merge_dicts(
-                combined.get(key, {}), domain_dict.get(key, {}), override
-            )
-
-        if duplicates:
-            duplicates = self.clean_duplicates(duplicates)
-            combined.update({"duplicates": duplicates})
-        return combined
-
-    @staticmethod
-    def extract_duplicates(list1: List[Any], list2: List[Any]) -> List[Any]:
-        """Extracts duplicates from two lists."""
-        if list1:
-            dict1 = {
-                (sorted(list(i.keys()))[0] if isinstance(i, dict) else i): i
-                for i in list1
-            }
-        else:
-            dict1 = {}
-
-        if list2:
-            dict2 = {
-                (sorted(list(i.keys()))[0] if isinstance(i, dict) else i): i
-                for i in list2
-            }
-        else:
-            dict2 = {}
-
-        set1 = set(dict1.keys())
-        set2 = set(dict2.keys())
-        dupes = set1.intersection(set2)
-        return sorted(list(dupes))
-
-    @staticmethod
-    def clean_duplicates(dupes: Dict[Text, Any]) -> Dict[Text, Any]:
-        """Removes keys for empty values."""
-        duplicates = dupes.copy()
-        for k in dupes:
-            if not dupes[k]:
-                duplicates.pop(k)
-
-        return duplicates
-
-    @staticmethod
-    def merge_dicts(
-        tempDict1: Dict[Text, Any],
-        tempDict2: Dict[Text, Any],
-        override_existing_values: bool = False,
-    ) -> Dict[Text, Any]:
-        """Merges two dicts."""
-        if override_existing_values:
-            merged_dicts, b = tempDict1.copy(), tempDict2.copy()
+        merged_dict = rasa.shared.utils.domain_loading.merge_domain_dicts(
+            domain2, domain1, override, is_dir
+        )
 
-        else:
-            merged_dicts, b = tempDict2.copy(), tempDict1.copy()
-        merged_dicts.update(b)
-        return merged_dicts
+        return merged_dict
 
-    @staticmethod
-    def merge_lists(list1: List[Any], list2: List[Any]) -> List[Any]:
-        """Merges two lists."""
-        return sorted(list(set(list1 + list2)))
-
-    @staticmethod
-    def merge_lists_of_dicts(
-        dict_list1: List[Dict],
-        dict_list2: List[Dict],
-        override_existing_values: bool = False,
-    ) -> List[Dict]:
-        """Merges two dict lists."""
-        dict1 = {
-            (sorted(list(i.keys()))[0] if isinstance(i, dict) else i): i
-            for i in dict_list1
-        }
-        dict2 = {
-            (sorted(list(i.keys()))[0] if isinstance(i, dict) else i): i
-            for i in dict_list2
-        }
-        merged_dicts = Domain.merge_dicts(dict1, dict2, override_existing_values)
-        return list(merged_dicts.values())
+    @property
+    def data(self) -> Dict:
+        """Returns original domain dict representation."""
+        self._data.update(self.as_dict())
```

Comment:
Why the `update`?

Reply:
The original domain dict could be missing mandatory keys such as `config` and `session_config` that get defaults during initialization - and the current utils function `merge_domain_dicts` expects that the domain that gets merged to the final/combined one has these config keys filled with defaults or otherwise. 

I will scrap using the `as_dict` and instead update only with the config values.

Reply:
@joejuzl Seems this change actually breaks tests(see [latest CI run](https://github.com/RasaHQ/rasa/runs/5188385918?check_suite_focus=true)). I'm thinking I should actually remove the `data` property from the Domain since it can't be used in merging either, wdyt?

Reply:
Just because it breaks tests doesn't mean it necessarily the wrong approach. Unit tests are often coupled somewhat to implementation - is there any specific _behaviour_ we can not replicate by using original data representation rather than converting back during `as_dict`?

Reply:
I've identified the problem that causes the retrieval intent related tests to fail - in the `ResponsesSyncImporter` code, the domain with retrieval intents that gets merged gets initialised with empty `data` dict (my bad!)

I'm trying to modify this `data` parameter to actually include intents and responses, however it's proving quite difficult since it seems the Domain does quite a few transformations in `as_dict` to handle these type of intents and responses. I'll continue working on this for 1-2 more hours 🤞🏼

Reply:
Ah found the trick! Setting override to `True` when merging the retrieval domain makes sure that the combined domain gets the data in the right format.